oscore-cf/src/main/java/org/eclipse/californium/oscore/RequestDecryptor.java

```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+ * Copyright (c) 2018 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Joakim Brorsson
+ *    Ludwig Seitz (RISE SICS)
+ *    Tobias Andersson (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import java.util.Arrays;
+import java.util.logging.Logger;
+
+import org.eclipse.californium.core.coap.OptionSet;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.CoAP.Code;
+import org.eclipse.californium.oscore.OSCoreCtx;
+import org.eclipse.californium.oscore.OSCoreCtxDB;
+import org.eclipse.californium.oscore.OSSerializer;
+import org.eclipse.californium.oscore.OptionJuggle;
+
+import com.upokecenter.cbor.CBORObject;
+import com.upokecenter.cbor.CBORType;
+
+import COSE.HeaderKeys;
+
+public class RequestDecryptor extends Decryptor {
+
+	/**
+	 * The logger
+	 */
+	private static final Logger LOGGER = Logger.getLogger(RequestDecryptor.class.getName());
+
+	/**
+	 * @param request the request to decrypt
+	 * @param db the database of OSCore contexts
+	 * 
+	 * @return the cid of the OSCore context use for this request
+	 * 
+	 * @throws OSException if decryption fails
+	 */
+	public static byte[] decrypt(Request request) throws OSException {
+
+		discardEOptions(request);
+
+		OSCoreCtxDB db = HashMapCtxDB.getInstance();
+		byte[] protectedData = collectData(request);
+		Encrypt0Message enc = decompression(protectedData, request);
+
+		CBORObject kid = enc.findAttribute(HeaderKeys.KID);
+		if (kid == null || !kid.getType().equals(CBORType.ByteString)) {
+			LOGGER.severe(OSExceptionHandler.MISSING_KID);
+			throw new OSException(OSExceptionHandler.MISSING_KID);
+		}
+		byte[] rid = kid.GetByteString();
+
+		OSCoreCtx ctx = db.getContext(rid);
+
+		if (ctx == null) {
+			LOGGER.severe(OSExceptionHandler.CONTEXT_NOT_FOUND);
+			throw new OSException(OSExceptionHandler.CONTEXT_NOT_FOUND);
+		}
+		
+		byte[] plaintext = new byte[0];
+		plaintext = decryptAndDecode(enc, request, ctx, null);
+		ctx.setCoAPCode(Code.valueOf(plaintext[0]));
+		plaintext = Arrays.copyOfRange(plaintext, 1, plaintext.length);
```

Comment:
I think, using a DatagramReader does a better job and solves also the broken "readPayload"!

Reply:
Even if that also clones the array :-).
Maybe we add an additional constructor to skip that clone.

Reply:
I will look into the DatagramReader.